File: skills/kazumi-rules-skill/scripts/kazumi_rule_codec.py
#!/usr/bin/env python3
"""Normalize Kazumi rule JSON and verify kazumi:// base64 payloads."""

from __future__ import annotations

import argparse
import base64
import binascii
import json
import re
import sys
from pathlib import Path
from typing import Any


CORE_FIELDS = (
    "api",
    "type",
    "name",
    "version",
    "baseURL",
    "searchURL",
    "searchList",
    "searchName",
    "searchResult",
    "chapterRoads",
    "chapterResult",
)

XPATH_FIELDS = (
    "searchList",
    "searchName",
    "searchResult",
    "chapterRoads",
    "chapterResult",
)

UNSUPPORTED_XPATH_PATTERNS = (
    (re.compile(r"\bcontains\s*\(", re.I), "contains() is not Kazumi-compatible; use [@attr*='value'] or [@attr~='value']"),
    (re.compile(r"\bstarts-with\s*\(", re.I), "starts-with() is not Kazumi-compatible; use [@attr^='value']"),
    (re.compile(r"\bnormalize-space\s*\(", re.I), "normalize-space() is not supported"),
    (re.compile(r"\bsubstring\s*\(", re.I), "substring() is not supported"),
    (re.compile(r"\bstring\s*\(", re.I), "string() is not supported"),
    (re.compile(r"\blast\s*\(", re.I), "last() is not supported"),
    (re.compile(r"\btext\s*\(", re.I), "text() predicates are not supported"),
    (re.compile(r"\band\b", re.I), "boolean and is not supported; choose one stable attribute"),
    (re.compile(r"\bor\b", re.I), "boolean or is not supported; choose one stable path"),
    (re.compile(r"\|"), "XPath union is not supported"),
    (re.compile(r"::"), "XPath axes such as following-sibling:: or ancestor:: are not supported"),
    (re.compile(r"(^|/)\.\.($|/)"), "parent traversal ../ is not supported"),
)


class RuleCodecError(ValueError):
    """Raised when a rule or encoded payload is invalid."""


def compact_json_bytes(rule: dict[str, Any]) -> bytes:
    return json.dumps(rule, ensure_ascii=False, separators=(",", ":"), sort_keys=False).encode("utf-8")


def pretty_json(rule: dict[str, Any]) -> str:
    return json.dumps(rule, ensure_ascii=False, indent=2)


def encode_rule(rule: dict[str, Any]) -> str:
    return base64.b64encode(compact_json_bytes(rule)).decode("ascii")


def import_link(rule: dict[str, Any]) -> str:
    return "kazumi://" + encode_rule(rule)


def decode_payload(payload: str) -> dict[str, Any]:
    raw = payload.strip()
    if raw.startswith("kazumi://"):
        raw = raw[len("kazumi://") :]
    raw = "".join(raw.split())

    padding = "=" * (-len(raw) % 4)
    try:
        decoded = base64.b64decode(raw + padding, validate=True)
    except binascii.Error as exc:
        raise RuleCodecError(f"base64 decode failed: {exc}") from exc

    try:
        value = json.loads(decoded.decode("utf-8"))
    except UnicodeDecodeError as exc:
        raise RuleCodecError("base64 payload is not UTF-8 JSON") from exc
    except json.JSONDecodeError as exc:
        raise RuleCodecError(f"base64 payload is not valid JSON: {exc}") from exc
    return require_rule_object(value)


def require_rule_object(value: Any) -> dict[str, Any]:
    if isinstance(value, list):
        raise RuleCodecError("Kazumi rule must be a single JSON object, not an array")
    if not isinstance(value, dict):
        raise RuleCodecError("Kazumi rule must decode to a JSON object")
    return value


def load_rule(source: str) -> dict[str, Any]:
    text = sys.stdin.read() if source == "-" else _read_source(source)
    stripped = text.strip()
    if not stripped:
        raise RuleCodecError("empty input")

    if stripped.startswith("kazumi://"):
        return decode_payload(stripped)

    if stripped.startswith("{") or stripped.startswith("["):
        try:
            return require_rule_object(json.loads(stripped))
        except json.JSONDecodeError as exc:
            raise RuleCodecError(f"JSON parse failed: {exc}") from exc

    return decode_payload(stripped)


def _read_source(source: str) -> str:
    path = Path(source)
    if path.exists():
        return path.read_text(encoding="utf-8")
    return source


def validate_rule(rule: dict[str, Any]) -> list[str]:
    errors: list[str] = []

    missing = [field for field in CORE_FIELDS if field not in rule]
    if missing:
        errors.append("missing required fields: " + ", ".join(missing))

    if str(rule.get("type", "")) != "anime":
        errors.append('type should be "anime"')

    search_url = str(rule.get("searchURL", ""))
    if "@keyword" not in search_url:
        errors.append("searchURL must contain @keyword")

    for field in XPATH_FIELDS:
        value = rule.get(field, "")
        if value is None:
            value = ""
        if not isinstance(value, str):
            errors.append(f"{field} must be a string")
            continue
        xpath = value.strip()
        if not xpath:
            continue
        if "@keyword" in xpath:
            errors.append(f"{field} must not contain @keyword")
        if not xpath.startswith("//"):
            errors.append(f"{field} should start with //")
        for pattern, message in UNSUPPORTED_XPATH_PATTERNS:
            if pattern.search(xpath):
                errors.append(f"{field}: {message}")

    encoded = encode_rule(rule)
    decoded = decode_payload(encoded)
    if decoded != rule:
        errors.append("base64 round-trip decoded JSON does not match the input rule")

    return errors


def build_report(rule: dict[str, Any]) -> dict[str, Any]:
    encoded = encode_rule(rule)
    link = "kazumi://" + encoded
    errors = validate_rule(rule)
    return {
        "ok": not errors,
        "errors": errors,
        "name": rule.get("name"),
        "api": rule.get("api"),
        "base64": encoded,
        "importLink": link,
        "decodedMatchesInput": decode_payload(encoded) == rule,
        "payloadShape": "object",
    }


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Normalize a single Kazumi rule object and verify its kazumi:// base64 encoding.",
    )
    parser.add_argument("input", help="Rule JSON file, raw JSON, raw base64, kazumi:// link, or - for stdin")
    parser.add_argument("--output", "-o", help="Write normalized JSON object to this file")
    parser.add_argument("--link-output", help="Write kazumi:// import link to this file")
    parser.add_argument("--compact", action="store_true", help="Print compact JSON instead of pretty JSON")
    parser.add_argument("--report", action="store_true", help="Print machine-readable validation report")
    parser.add_argument("--quiet", action="store_true", help="Only print errors")
    args = parser.parse_args(argv)

    try:
        rule = load_rule(args.input)
        report = build_report(rule)
    except RuleCodecError as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        return 2

    json_text = (
        json.dumps(rule, ensure_ascii=False, separators=(",", ":"))
        if args.compact
        else pretty_json(rule)
    )
    link = report["importLink"]

    if args.output:
        Path(args.output).write_text(json_text + "\n", encoding="utf-8")
    if args.link_output:
        Path(args.link_output).write_text(link + "\n", encoding="utf-8")

    if report["errors"]:
        for error in report["errors"]:
            print(f"ERROR: {error}", file=sys.stderr)

    if not args.quiet:
        if args.report:
            print(json.dumps(report, ensure_ascii=False, indent=2))
        else:
            print("Normalized rule JSON:")
            print(json_text)
            print()
            print("Import link:")
            print(link)

    return 0 if report["ok"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/kazumi-rules-skill/scripts/kazumi_rule_codec.ts
#!/usr/bin/env node
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

export type Rule = Record<string, any>;
export const SECRET_RE =
  /authorization|cookie|token|secret|api[-_]?key|csrf|signature/i;
export const TEMPLATE_RE = /(?<![A-Za-z0-9_])@([A-Za-z_][A-Za-z0-9_]*)/g;

const COMMON_FIELDS = ["api", "type", "name", "version", "baseURL"];
const SEARCH_XPATH_FIELDS = [
  "searchURL",
  "searchList",
  "searchName",
  "searchResult",
];
const CHAPTER_XPATH_FIELDS = ["chapterRoads", "chapterResult"];
const UNSUPPORTED_XPATH: Array<[RegExp, string]> = [
  [
    /\bcontains\s*\(/i,
    "contains() is not Kazumi-compatible; use [@attr*='value']",
  ],
  [
    /\bstarts-with\s*\(/i,
    "starts-with() is not Kazumi-compatible; use [@attr^='value']",
  ],
  [
    /\b(?:normalize-space|substring|string|last|text)\s*\(/i,
    "XPath functions/text() are not supported",
  ],
  [/\b(?:and|or)\b/i, "boolean predicates are not supported"],
  [/\|/, "XPath union is not supported"],
  [/::/, "XPath axes are not supported"],
  [/(^|\/)\.\.($|\/)/, "parent traversal ../ is not supported"],
];

export class RuleCodecError extends Error {}

export function compactJson(rule: Rule): string {
  return JSON.stringify(rule);
}

export function encodeRule(rule: Rule): string {
  return Buffer.from(compactJson(rule), "utf8").toString("base64");
}

export function importLink(rule: Rule): string {
  return `kazumi://${encodeRule(rule)}`;
}

export function requireRuleObject(value: unknown): Rule {
  if (Array.isArray(value))
    throw new RuleCodecError(
      "Kazumi rule must be a single JSON object, not an array",
    );
  if (value === null || typeof value !== "object")
    throw new RuleCodecError("Kazumi rule must decode to a JSON object");
  return value as Rule;
}

export function decodePayload(input: string): Rule {
  let raw = input.trim();
  const scheme = raw.match(/^kazumi:(?:\/\/)?/i);
  if (scheme) raw = raw.slice(scheme[0].length);
  else if (raw.includes("://"))
    throw new RuleCodecError("invalid Kazumi rule link scheme");
  try {
    raw = decodeURIComponent(raw);
  } catch (error) {
    throw new RuleCodecError(`percent decoding failed: ${String(error)}`);
  }
  raw = raw.replace(/\s/g, "").replace(/-/g, "+").replace(/_/g, "/");
  if (!raw) throw new RuleCodecError("Kazumi rule link is empty");
  if (!/^[A-Za-z0-9+/]*={0,2}$/.test(raw) || raw.length % 4 === 1)
    throw new RuleCodecError("base64 decode failed");
  raw += "=".repeat((4 - (raw.length % 4)) % 4);
  try {
    return requireRuleObject(
      JSON.parse(Buffer.from(raw, "base64").toString("utf8")),
    );
  } catch (error) {
    if (error instanceof RuleCodecError) throw error;
    throw new RuleCodecError(
      `base64 payload is not valid UTF-8 JSON: ${String(error)}`,
    );
  }
}

export function loadRule(source: string): Rule {
  const text =
    source === "-"
      ? readFileSync(0, "utf8")
      : existsSync(source)
        ? readFileSync(source, "utf8")
        : source;
  const trimmed = text.trim();
  if (!trimmed) throw new RuleCodecError("empty input");
  if (trimmed.startsWith("{") || trimmed.startsWith("[")) {
    try {
      return requireRuleObject(JSON.parse(trimmed));
    } catch (error) {
      if (error instanceof RuleCodecError) throw error;
      throw new RuleCodecError(`JSON parse failed: ${String(error)}`);
    }
  }
  return decodePayload(trimmed);
}

export function ruleMode(rule: Rule, phase: "search" | "chapter"): string {
  return String(rule[`${phase}Mode`] ?? "xpath");
}

function bracketEnd(expression: string, start: number): number {
  let quote = "";
  let escaped = false;
  for (let index = start + 1; index < expression.length; index++) {
    const char = expression[index];
    if (escaped) escaped = false;
    else if (char === "\\") escaped = true;
    else if (quote) {
      if (char === quote) quote = "";
    } else if (char === "'" || char === '"') quote = char;
    else if (char === "]") return index;
  }
  return -1;
}

export function validateJsonPath(expression: unknown): string | null {
  if (typeof expression !== "string" || !expression.startsWith("$"))
    return "JSONPath must be a string starting with $";
  let index = 1;
  while (index < expression.length) {
    if (expression[index] === ".") {
      const match = expression.slice(++index).match(/^[A-Za-z0-9_$-]+/);
      if (!match) return `unsupported JSONPath: ${expression}`;
      index += match[0].length;
      continue;
    }
    if (expression[index] === "[") {
      const end = bracketEnd(expression, index);
      if (end < 0) return `JSONPath is missing ]: ${expression}`;
      const part = expression.slice(index + 1, end).trim();
      const quoted =
        part.length >= 2 &&
        ["'", '"'].includes(part[0]) &&
        part.at(-1) === part[0];
      if (!(part === "*" || /^\d+$/.test(part) || quoted))
        return `unsupported JSONPath segment [${part}]`;
      index = end + 1;
      continue;
    }
    return `unsupported JSONPath: ${expression}`;
  }
  return null;
}

export function templateNames(value: unknown): Set<string> {
  const names = new Set<string>();
  const visit = (item: unknown): void => {
    if (typeof item === "string")
      for (const match of item.matchAll(TEMPLATE_RE)) names.add(match[1]);
    else if (Array.isArray(item)) item.forEach(visit);
    else if (item && typeof item === "object")
      for (const [key, child] of Object.entries(item)) {
        visit(key);
        visit(child);
      }
  };
  visit(value);
  return names;
}

function validateXpath(rule: Rule, fields: string[], errors: string[]): void {
  for (const field of fields) {
    const value = rule[field];
    if (typeof value !== "string" || !value.trim()) {
      errors.push(`${field} must be a non-empty string in XPath mode`);
      continue;
    }
    if (field === "searchURL") {
      if (!value.includes("@keyword"))
        errors.push("searchURL must contain @keyword in XPath mode");
      continue;
    }
    if (!value.startsWith("//")) errors.push(`${field} should start with //`);
    if (value.includes("@keyword"))
      errors.push(`${field} must not contain @keyword`);
    for (const match of value.matchAll(/\[([^\]]+)\]/g)) {
      const predicate = match[1];
      if (
        !/^\s*\d+\s*$/.test(predicate) &&
        !/^\s*@[\w:-]+\s*(?:=|~=|\^=|\$=|\*=)\s*(['"]).*\1\s*$/.test(predicate)
      )
        errors.push(
          `${field}: unsupported or unquoted predicate [${predicate}]`,
        );
    }
    for (const [pattern, message] of UNSUPPORTED_XPATH)
      if (pattern.test(value)) errors.push(`${field}: ${message}`);
  }
}

function validateRequest(
  config: unknown,
  label: string,
  allowed: Set<string>,
  errors: string[],
): void {
  if (!config || typeof config !== "object" || Array.isArray(config)) {
    errors.push(`${label} must be an object`);
    return;
  }
  const request = config as Rule;
  const method = String(request.method ?? "GET").toUpperCase();
  if (!new Set(["GET", "POST"]).has(method))
    errors.push(`${label}.method must be GET or POST`);
  if (typeof request.url !== "string" || !request.url.trim())
    errors.push(`${label}.url must be a non-empty string`);
  for (const key of ["headers", "query"])
    if (
      key in request &&
      (!request[key] ||
        typeof request[key] !== "object" ||
        Array.isArray(request[key]))
    )
      errors.push(`${label}.${key} must be an object`);
  const bodyType = String(request.bodyType ?? "none");
  if (!new Set(["none", "json", "form"]).has(bodyType))
    errors.push(`${label}.bodyType must be none, json, or form`);
  if (method === "POST" && bodyType !== "none" && !("body" in request))
    errors.push(
      `${label}.body is required for active POST ${bodyType} requests`,
    );
  const unknown = [...templateNames(request)].filter(
    (name) => !allowed.has(name),
  );
  if (unknown.length)
    errors.push(
      `${label} uses unknown template variables: ${unknown
        .sort()
        .map((name) => `@${name}`)
        .join(", ")}`,
    );
}

function validateApiSearch(rule: Rule, errors: string[]): void {
  const config = rule.searchApiConfig;
  if (!config || typeof config !== "object" || Array.isArray(config)) {
    errors.push("searchApiConfig must be an object in API mode");
    return;
  }
  validateRequest(
    config.request,
    "searchApiConfig.request",
    new Set(["keyword"]),
    errors,
  );
  for (const field of ["listPath", "namePath", "sourcePath"]) {
    const error = validateJsonPath(config[field]);
    if (error) errors.push(`searchApiConfig.${field}: ${error}`);
  }
}

function validateApiChapters(rule: Rule, errors: string[]): void {
  const config = rule.chapterApiConfig;
  if (!config || typeof config !== "object" || Array.isArray(config)) {
    errors.push("chapterApiConfig must be an object in API mode");
    return;
  }
  validateRequest(
    config.request,
    "chapterApiConfig.request",
    new Set(["source"]),
    errors,
  );
  const format = String(config.format ?? "nested");
  if (!new Set(["nested", "delimited"]).has(format)) {
    errors.push("chapterApiConfig.format must be nested or delimited");
    return;
  }
  const variables = config.variables ?? {};
  if (!variables || typeof variables !== "object" || Array.isArray(variables))
    errors.push("chapterApiConfig.variables must be an object");
  else
    for (const [name, path] of Object.entries(variables)) {
      const error = validateJsonPath(path);
      if (error) errors.push(`chapterApiConfig.variables.${name}: ${error}`);
    }
  if (format === "delimited") {
    for (const field of ["roadNamesPath", "roadEpisodesPath"]) {
      const error = validateJsonPath(config[field]);
      if (error) errors.push(`chapterApiConfig.${field}: ${error}`);
    }
    for (const [field, fallback] of [
      ["roadSeparator", "$$$"],
      ["episodeSeparator", "#"],
      ["fieldSeparator", "$"],
    ])
      if (!String(config[field] ?? fallback))
        errors.push(`chapterApiConfig.${field} must not be empty`);
  } else {
    for (const field of ["roadsPath", "roadNamePath"])
      if (config[field]) {
        const error = validateJsonPath(config[field]);
        if (error) errors.push(`chapterApiConfig.${field}: ${error}`);
      }
    for (const field of ["episodesPath", "episodeNamePath"]) {
      const error = validateJsonPath(config[field]);
      if (error) errors.push(`chapterApiConfig.${field}: ${error}`);
    }
    if (config.episodeUrlPath) {
      const error = validateJsonPath(config.episodeUrlPath);
      if (error) errors.push(`chapterApiConfig.episodeUrlPath: ${error}`);
    } else if (!config.episodePage || typeof config.episodePage !== "object")
      errors.push("chapterApiConfig requires episodeUrlPath or episodePage");
  }
  if (config.episodePage !== undefined) {
    const page = config.episodePage;
    if (
      !page ||
      typeof page !== "object" ||
      typeof page.url !== "string" ||
      !page.url.trim()
    )
      errors.push(
        "chapterApiConfig.episodePage.url must be a non-empty string",
      );
    else if (
      page.query !== undefined &&
      (!page.query ||
        typeof page.query !== "object" ||
        Array.isArray(page.query))
    )
      errors.push("chapterApiConfig.episodePage.query must be an object");
    const allowed = new Set([
      "source",
      "episodeUrl",
      "roadIndex",
      "roadNumber",
      "episodeIndex",
      "episodeNumber",
      ...Object.keys(variables),
    ]);
    const unknown = [...templateNames(page)].filter(
      (name) => !allowed.has(name),
    );
    if (unknown.length)
      errors.push(
        `chapterApiConfig.episodePage uses unknown template variables: ${unknown
          .sort()
          .map((name) => `@${name}`)
          .join(", ")}`,
      );
  }
}

export function validateRule(rule: Rule): string[] {
  const errors: string[] = [];
  const missing = COMMON_FIELDS.filter((field) => !(field in rule));
  if (missing.length)
    errors.push(`missing required fields: ${missing.join(", ")}`);
  if (String(rule.type ?? "anime") !== "anime")
    errors.push('type should be "anime"');
  const search = ruleMode(rule, "search");
  const chapter = ruleMode(rule, "chapter");
  if (!new Set(["xpath", "api"]).has(search))
    errors.push("searchMode must be xpath or api");
  if (!new Set(["xpath", "api"]).has(chapter))
    errors.push("chapterMode must be xpath or api");
  if (
    [search, chapter].includes("api") &&
    (!/^\d+$/.test(String(rule.api)) || Number(rule.api) < 8)
  )
    errors.push("API search/chapter mode requires api level 8 or newer");
  if (search === "xpath") validateXpath(rule, SEARCH_XPATH_FIELDS, errors);
  else if (search === "api") validateApiSearch(rule, errors);
  if (chapter === "xpath") validateXpath(rule, CHAPTER_XPATH_FIELDS, errors);
  else if (chapter === "api") validateApiChapters(rule, errors);
  if (JSON.stringify(decodePayload(encodeRule(rule))) !== JSON.stringify(rule))
    errors.push("base64 round-trip decoded JSON does not match the input rule");
  return errors;
}

export function buildReport(rule: Rule): Rule {
  const base64 = encodeRule(rule);
  const errors = validateRule(rule);
  return {
    ok: !errors.length,
    errors,
    name: rule.name,
    api: rule.api,
    searchMode: ruleMode(rule, "search"),
    chapterMode: ruleMode(rule, "chapter"),
    base64,
    importLink: `kazumi://${base64}`,
    decodedMatchesInput:
      JSON.stringify(decodePayload(base64)) === JSON.stringify(rule),
    payloadShape: "object",
  };
}

function usage(): never {
  console.error(
    "Usage: node kazumi_rule_codec.ts INPUT [--output FILE] [--link-output FILE] [--compact] [--report] [--quiet]",
  );
  process.exit(2);
}

export function main(argv = process.argv.slice(2)): number {
  if (!argv.length || argv.includes("--help")) {
    if (argv.includes("--help")) {
      console.log("Validate and encode one Kazumi XPath/API rule object.");
      return 0;
    }
    usage();
  }
  const input = argv[0];
  const option = (name: string): string | undefined => {
    const index = argv.indexOf(name);
    return index >= 0 ? argv[index + 1] : undefined;
  };
  try {
    const rule = loadRule(input);
    const report = buildReport(rule);
    const text = argv.includes("--compact")
      ? JSON.stringify(rule)
      : JSON.stringify(rule, null, 2);
    const output = option("--output") ?? option("-o");
    if (output) writeFileSync(output, `${text}\n`, "utf8");
    const linkOutput = option("--link-output");
    if (linkOutput) writeFileSync(linkOutput, `${report.importLink}\n`, "utf8");
    for (const error of report.errors) console.error(`ERROR: ${error}`);
    if (!argv.includes("--quiet"))
      console.log(
        argv.includes("--report")
          ? JSON.stringify(report, null, 2)
          : `Normalized rule JSON:\n${text}\n\nImport link:\n${report.importLink}`,
      );
    return report.ok ? 0 : 1;
  } catch (error) {
    console.error(
      `ERROR: ${error instanceof Error ? error.message : String(error)}`,
    );
    return 2;
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href)
  process.exitCode = main();
